File: ios/MenuMakerUITests/PageObjects/NotificationPage.swift
//
//  NotificationPage.swift
//  MenuMakerUITests
//
//  Page Object for Notifications Screen
//

import XCTest

struct NotificationPage {
    let app: XCUIApplication

    // MARK: - Elements

    var notificationsList: XCUIElementQuery {
        app.scrollViews.otherElements.matching(identifier: "NotificationItem")
    }

    var firstNotification: XCUIElement {
        notificationsList.firstMatch
    }

    var unreadNotifications: XCUIElementQuery {
        app.scrollViews.otherElements.matching(NSPredicate(format: "identifier CONTAINS 'Notification' AND label CONTAINS '•'"))
    }

    var emptyStateMessage: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS[c] 'no notification'")).firstMatch
    }

    var markAllReadButton: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'mark all' OR label CONTAINS[c] 'read all'")).firstMatch
    }

    var filterButtons: XCUIElementQuery {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'all' OR label CONTAINS[c] 'unread' OR label CONTAINS[c] 'orders' OR label CONTAINS[c] 'updates'"))
    }

    var allFilter: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'all'")).firstMatch
    }

    var unreadFilter: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'unread'")).firstMatch
    }

    var ordersFilter: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'order'")).firstMatch
    }

    var clearAllButton: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'clear' OR label CONTAINS[c] 'delete all'")).firstMatch
    }

    var settingsButton: XCUIElement {
        app.navigationBars.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'setting' OR label CONTAINS '⚙️'")).firstMatch
    }

    var backButton: XCUIElement {
        app.navigationBars.buttons.firstMatch
    }

    // Notification types
    var orderNotifications: XCUIElementQuery {
        app.scrollViews.otherElements.matching(NSPredicate(format: "label CONTAINS[c] 'order'"))
    }

    var deliveryNotifications: XCUIElementQuery {
        app.scrollViews.otherElements.matching(NSPredicate(format: "label CONTAINS[c] 'delivery' OR label CONTAINS[c] 'delivered'"))
    }

    var promoNotifications: XCUIElementQuery {
        app.scrollViews.otherElements.matching(NSPredicate(format: "label CONTAINS[c] 'offer' OR label CONTAINS[c] 'discount'"))
    }

    var notificationBadge: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label MATCHES '\\\\d+' AND identifier CONTAINS 'badge'")).firstMatch
    }

    // MARK: - Actions

    @discardableResult
    func tapFirstNotification() -> NotificationPage {
        firstNotification.tap()
        sleep(1)
        return self
    }

    @discardableResult
    func tapNotification(at index: Int) -> NotificationPage {
        let notification = notificationsList.element(boundBy: index)
        if notification.waitForExistence(timeout: 2) {
            notification.tap()
            sleep(1)
        }
        return self
    }

    @discardableResult
    func swipeToDeleteNotification(at index: Int = 0) -> NotificationPage {
        let notification = notificationsList.element(boundBy: index)
        if notification.exists {
            notification.swipeLeft()
            let deleteButton = app.buttons["Delete"]
            if deleteButton.waitForExistence(timeout: 1) {
                deleteButton.tap()
                sleep(1)
            }
        }
        return self
    }

    @discardableResult
    func markAllAsRead() -> NotificationPage {
        if markAllReadButton.waitForExistence(timeout: 2) {
            markAllReadButton.tap()
            sleep(1)
        }
        return self
    }

    @discardableResult
    func clearAllNotifications() -> NotificationPage {
        if clearAllButton.waitForExistence(timeout: 2) {
            clearAllButton.tap()
            sleep(1)

            // Confirm if needed
            let confirmButton = app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'confirm' OR label CONTAINS[c] 'yes'")).firstMatch
            if confirmButton.waitForExistence(timeout: 1) {
                confirmButton.tap()
                sleep(1)
            }
        }
        return self
    }

    @discardableResult
    func filterByAll() -> NotificationPage {
        allFilter.tap()
        sleep(1)
        return self
    }

    @discardableResult
    func filterByUnread() -> NotificationPage {
        unreadFilter.tap()
        sleep(1)
        return self
    }

    @discardableResult
    func filterByOrders() -> NotificationPage {
        ordersFilter.tap()
        sleep(1)
        return self
    }

    @discardableResult
    func pullToRefresh() -> NotificationPage {
        let scrollView = app.scrollViews.firstMatch
        if scrollView.exists {
            let start = scrollView.coordinate(withNormalizedOffset: CGVector(dx: 0.5, dy: 0.2))
            let end = scrollView.coordinate(withNormalizedOffset: CGVector(dx: 0.5, dy: 0.8))
            start.press(forDuration: 0, thenDragTo: end)
            sleep(1)
        }
        return self
    }

    @discardableResult
    func tapSettings() -> NotificationPage {
        settingsButton.tap()
        sleep(1)
        return self
    }

    @discardableResult
    func goBack() -> NotificationPage {
        backButton.tap()
        return self
    }

    // MARK: - Assertions

    @discardableResult
    func assertScreenDisplayed(timeout: TimeInterval = 2) -> NotificationPage {
        XCTAssertTrue(firstNotification.waitForExistence(timeout: timeout) ||
                     emptyStateMessage.waitForExistence(timeout: timeout) ||
                     markAllReadButton.waitForExistence(timeout: timeout),
                     "Notifications screen should be displayed")
        return self
    }

    @discardableResult
    func assertNotificationsDisplayed() -> NotificationPage {
        XCTAssertTrue(firstNotification.exists, "Notifications should be displayed")
        return self
    }

    @discardableResult
    func assertEmptyState() -> NotificationPage {
        XCTAssertTrue(emptyStateMessage.exists, "Empty state should be displayed")
        return self
    }

    @discardableResult
    func assertUnreadNotificationsExist() -> NotificationPage {
        XCTAssertGreaterThan(unreadNotifications.count, 0, "Unread notifications should exist")
        return self
    }

    @discardableResult
    func assertNotificationCount(_ expectedCount: Int) -> NotificationPage {
        let actualCount = notificationsList.count
        XCTAssertEqual(actualCount, expectedCount,
                      "Should have \(expectedCount) notifications, found \(actualCount)")
        return self
    }

    @discardableResult
    func assertOrderNotificationsExist() -> NotificationPage {
        XCTAssertGreaterThan(orderNotifications.count, 0, "Order notifications should exist")
        return self
    }

    @discardableResult
    func assertFilterOptionsDisplayed() -> NotificationPage {
        XCTAssertGreaterThan(filterButtons.count, 0, "Filter options should be displayed")
        return self
    }

    @discardableResult
    func assertMarkAllReadButtonVisible() -> NotificationPage {
        XCTAssertTrue(markAllReadButton.exists, "Mark all read button should be visible")
        return self
    }

    @discardableResult
    func assertNotificationBadgeCount(_ count: Int) -> NotificationPage {
        if count > 0 {
            XCTAssertTrue(notificationBadge.exists, "Notification badge should be visible")
            let badgeText = notificationBadge.label
            if let badgeCount = Int(badgeText) {
                XCTAssertEqual(badgeCount, count, "Badge should show \(count) notifications")
            }
        } else {
            XCTAssertFalse(notificationBadge.exists, "Notification badge should not be visible")
        }
        return self
    }
}

File: android/app/src/androidTest/kotlin/com/menumaker/pageobjects/NotificationPage.kt
package com.menumaker.pageobjects

import androidx.compose.ui.test.*
import androidx.compose.ui.test.junit4.ComposeTestRule

/**
 * Page Object for Notifications Screen
 * Provides fluent API for notification interactions
 */
class NotificationPage(private val composeTestRule: ComposeTestRule) {

    // Elements
    private val notificationsList = composeTestRule.onAllNodesWithTag("NotificationItem")
    private val unreadNotifications = composeTestRule.onAllNodes(
        hasTestTag("NotificationItem") and hasText("•", substring = true)
    )
    private val emptyStateMessage = composeTestRule.onNodeWithText("no notification", substring = true, ignoreCase = true)
    private val markAllReadButton = composeTestRule.onNode(
        hasText("mark all", substring = true, ignoreCase = true) or
        hasText("read all", substring = true, ignoreCase = true)
    )
    private val clearAllButton = composeTestRule.onNode(
        hasText("clear", substring = true, ignoreCase = true) or
        hasText("delete all", substring = true, ignoreCase = true)
    )
    private val settingsButton = composeTestRule.onNodeWithContentDescription("Settings", substring = true)
    private val backButton = composeTestRule.onNodeWithContentDescription("Back", substring = true)

    // Filter tabs
    private val allFilter = composeTestRule.onNodeWithText("all", ignoreCase = true)
    private val unreadFilter = composeTestRule.onNodeWithText("unread", ignoreCase = true)
    private val ordersFilter = composeTestRule.onNodeWithText("order", substring = true, ignoreCase = true)

    // Actions
    fun tapFirstNotification(): NotificationPage {
        if (notificationsList.fetchSemanticsNodes().isNotEmpty()) {
            notificationsList[0].performClick()
            Thread.sleep(1000)
        }
        return this
    }

    fun tapNotification(index: Int): NotificationPage {
        if (notificationsList.fetchSemanticsNodes().size > index) {
            notificationsList[index].performClick()
            Thread.sleep(1000)
        }
        return this
    }

    fun swipeToDeleteNotification(index: Int = 0): NotificationPage {
        if (notificationsList.fetchSemanticsNodes().size > index) {
            notificationsList[index].performTouchInput {
                swipeLeft()
            }
            val deleteButton = composeTestRule.onNodeWithText("Delete", ignoreCase = true)
            if (deleteButton.fetchSemanticsNode(false) != null) {
                deleteButton.performClick()
                Thread.sleep(1000)
            }
        }
        return this
    }

    fun markAllAsRead(): NotificationPage {
        markAllReadButton.performClick()
        Thread.sleep(1000)
        return this
    }

    fun clearAllNotifications(): NotificationPage {
        clearAllButton.performClick()
        Thread.sleep(1000)

        // Confirm if dialog appears
        val confirmButton = composeTestRule.onNode(
            hasText("confirm", substring = true, ignoreCase = true) or
            hasText("yes", ignoreCase = true)
        )
        if (confirmButton.fetchSemanticsNode(false) != null) {
            confirmButton.performClick()
            Thread.sleep(1000)
        }
        return this
    }

    fun filterByAll(): NotificationPage {
        allFilter.performClick()
        Thread.sleep(1000)
        return this
    }

    fun filterByUnread(): NotificationPage {
        unreadFilter.performClick()
        Thread.sleep(1000)
        return this
    }

    fun filterByOrders(): NotificationPage {
        ordersFilter.performClick()
        Thread.sleep(1000)
        return this
    }

    fun tapSettings(): NotificationPage {
        settingsButton.performClick()
        Thread.sleep(1000)
        return this
    }

    fun goBack(): NotificationPage {
        backButton.performClick()
        return this
    }

    // Assertions
    fun assertScreenDisplayed(): NotificationPage {
        composeTestRule.waitUntil(timeoutMillis = 2000) {
            notificationsList.fetchSemanticsNodes().isNotEmpty() ||
            emptyStateMessage.fetchSemanticsNode(false) != null ||
            markAllReadButton.fetchSemanticsNode(false) != null
        }
        return this
    }

    fun assertNotificationsDisplayed(): NotificationPage {
        assert(notificationsList.fetchSemanticsNodes().isNotEmpty()) {
            "Notifications should be displayed"
        }
        return this
    }

    fun assertEmptyState(): NotificationPage {
        emptyStateMessage.assertExists()
        return this
    }

    fun assertUnreadNotificationsExist(): NotificationPage {
        assert(unreadNotifications.fetchSemanticsNodes().isNotEmpty()) {
            "Unread notifications should exist"
        }
        return this
    }

    fun assertNotificationCount(expectedCount: Int): NotificationPage {
        val actualCount = notificationsList.fetchSemanticsNodes().size
        assert(actualCount == expectedCount) {
            "Should have $expectedCount notifications, found $actualCount"
        }
        return this
    }

    fun assertOrderNotificationsExist(): NotificationPage {
        val orderNotifs = composeTestRule.onAllNodes(
            hasText("order", substring = true, ignoreCase = true)
        )
        assert(orderNotifs.fetchSemanticsNodes().isNotEmpty()) {
            "Order notifications should exist"
        }
        return this
    }

    fun assertFilterOptionsDisplayed(): NotificationPage {
        allFilter.assertExists()
        return this
    }

    fun assertMarkAllReadButtonVisible(): NotificationPage {
        markAllReadButton.assertExists()
        return this
    }

    fun assertNotificationBadgeCount(count: Int): NotificationPage {
        val badge = composeTestRule.onNodeWithTag("notification-badge")
        if (count > 0) {
            badge.assertExists()
        } else {
            badge.assertDoesNotExist()
        }
        return this
    }
}
